### PR TITLE
RD-1579 Modify the CLI to support complex filters

### PR DESCRIPTION
**Pull request by YoniItzhak**

### CODE REVIEW

File: dev-requirements.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
 wheel==0.29.0
 setuptools==40.7.3
 virtualenv==15.1.0
-https://github.com/cloudify-cosmo/cloudify-common/archive/master.zip#egg=cloudify-common[dispatcher]
+https://github.com/cloudify-cosmo/cloudify-common/archive/RD-1472-complex-filters.zip#egg=cloudify-common[dispatcher]
```
Comment (glukhman) [inline dev-requirements.txt:4]: reminder when merging to master

**Review by glukhman — APPROVED** [review]

File: cloudify_cli/commands/deployments.py [hunk 2]
```diff
@@ -186,7 +188,9 @@ def _print_single_update(deployment_update_dict,
 @cfy.command(name='list', short_help='List deployments [manager only]')
 @cfy.options.blueprint_id()
 @click.option('--group-id', '-g')
-@cfy.options.filter_rules
+@cfy.options.filter_id
+@cfy.options.labels_filter
+@cfy.options.attrs_filter('deployment')
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/commands/deployments.py:193]: Great comment! I did the same for `labels_rules` and `attrs_rules`.

File: cloudify_cli/filters_utils.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+        super(CloudifyCliError, self).__init__(
+            'The labels filter rule `{0}` is not in the right format. It must '
+            'be one of: <key>=<value>, <key>!=<value>, <key> is null, '
+            '<key> is not null. <value> can be a single string or a list of '
+            'strings of the form [<value1>,<value2>,...]'.format(
+                labels_filter_value)
+        )
+
+
+class BadAttributesFilterRule(CloudifyCliError):
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/filters_utils.py:44]: Right, changed to `InvalidLabelsFilterRuleFormat` and `InvalidAttributesFilterRuleFormat`.

File: cloudify_cli/filters_utils.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    validate_visibility(visibility)
+
+    filter_rules = get_filter_rules(labels_rules, attrs_rules)
+    try:
+        new_filter = filters_client.create(filter_id, filter_rules, visibility)
+    except InvalidFilterRule as e:
+        e._message = _modify_err_msg(e.err_filter_rule, e.err_reason)
+        raise e
+
+    logger.info("{0}' filter `{1}` created".format(resource_name.capitalize(),
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/filters_utils.py:95]: I am really ashamed I missed that, again.

File: cloudify_cli/filters_utils.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+                key = match_not_empty.group(1).lower()
+                attrs_filter_rules.append(
+                    AttrsFilterRule(key, [], 'is not empty'))
+            else:
+                raise BadAttributesFilterRule(attrs_filter)
+
+    return attrs_filter_rules
+
+
+def _parse_filter_rule(filter_rule, sign, filter_rule_err_class):
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/filters_utils.py:250]: You're absolutely right. I modified the logic to use the `FilterRule` class better.

File: cloudify_cli/filters_utils.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+
+def _get_label_value(raw_label_value):
+    if raw_label_value.startswith('[') and raw_label_value.endswith(']'):
+        return raw_label_value.strip('[]').split(',')
+
+    return [raw_label_value]
+
+
+def get_filter_rules(labels_filter_rules, attrs_filter_rules):
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/filters_utils.py:284]: The only issue is that `labels_filte_rules` and `attrs_filter_rules` can be `None`.

File: cloudify_cli/filters_utils.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+            if len(filter_rule['values']) == 1:
+                filter_rule_str = '{key}{operator}{values}'.format(
+                    key=filter_rule['key'], operator=cli_operator,
+                    values=filter_rule['values'][0])
+            else:
+                filter_rule_str = '{key}{operator}[{values}]'.format(
+                    key=filter_rule['key'], operator=cli_operator,
+                    values=','.join(filter_rule['values']))
+
+        filter_rules_str_list.append(filter_rule_str)
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/filters_utils.py:321]: Done. Replaced to use `__str__`.

File: cloudify_cli/filters_utils.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+    return '\"{0}\"'.format(' and '.join(filter_rules_str_list))
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/filters_utils.py:323]: Right..

File: cloudify_cli/table.py [hunk 9]
```diff
@@ -44,7 +44,7 @@ def generate(cols, data, defaults=None, labels=None):
                    key's that don't exist in the data itself.
 
                    for example: {'deploymentId':'123'} will set the
-                   deploymentId value for all rows to '123'.
+                   deploymentId va lue for all rows to '123'.
```
[the remark this answers is not in the document]
Comment (YoniItzhak) [reply cloudify_cli/table.py]: I have no idea how that happened.

**Review by tehasdf — APPROVED** [review]

**Review by glukhman — APPROVED** [review]

**Pull request merged**